doc/designs/quic-design/error-handling.md

```diff
@@ -0,0 +1,88 @@
+Error handling in QUIC code
+===========================
+
+Current situation with TLS
+--------------------------
+
+The errors are put on the error stack (rather a queue but error stack is
+used throughout the code base) during the libssl API calls. In most
+(if not all) cases they should appear there only if the API call returns an
+error return value. The `SSL_get_error()` call depends on the stack being
+clean before the API call to be properly able to determine if the API
+call caused a library or system (I/O) error.
+
+As the error stacks are thread-local calls from separate threads (even with
+the same SSL object) push errors to these separate error stacks. This is
+not a problem as applications are supposed to check for errors immediately
+after the API call on the same thread. There is no such thing as
+Thread-assisted mode of operation.
+
+Constraints
+-----------
+
+We need to keep using the existing ERR API as doing otherwise would
+complicate the existing applications and break our API compatibility promise.
+Even the ERR_STATE structure is public, although deprecated, an thus its
+structure and semantics cannot be changed.
+
+The error stack access is not under a lock (because it is thread-local).
+This complicates _moving errors between threads_.
+
+Error stack entries contain allocated data, copying entries between threads
+would mean duplicating it or losing it.
+
+Assumptions
+-----------
+
+This document assumes the error state of the QUIC connection (or stream for
+stream level errors) is handled separately.
+
+We can assume the internal assistance thread is well-behaving in regards
+to the error stack.
+
+We assume there are two types of errors that can be raised in the QUIC
+library calls and in the subordinate libcrypto (and provider) calls. First
+type is an intermittent error that does not really affect the state of the
+QUIC connection - for example EAGAIN returned on a syscall, or unavailability
+of some algorithm where there are other algorithms to try. Second type
+is a permanent error that affect the error state of the QUIC connection.
+
+Design
+------
+
+Return value of SSL_get_error() on QUIC connections or streams does not
+depend on the error stack contents.
+
+Intermittent errors are handled within the library and cleared from the
+error stack before returning to the user.
+
+Permanent errors happenning within the assist thread need to be transferred
+to the regular user thread. To simplify the implementation they are required
+to appear on the error stack reachable by the application ERR_ calls only
+after the SSL_get_error() is called from the application.
```

Comment:
It's not just the assist thread we have to worry about. How does SSL_tick() work in relation to errors? Does it fail immediately? Or does the failure occur on the next io function call? We have to keep in mind that a single SSL_tick() call can, when we have server functionality, be used to process packets from multiple connections. In 3.2 we will have multiple SSL objects on different threads. So an SSL_tick() call on one thread could encounter an error related to a stream being processed on a different thread.

What happens with an error that affects multiple SSL objects? E.g. where there is some fundamental error on a connection that leads to all related stream objects to abort.

Reply:
The current SSL_tick() API has SSL object as a parameter. That SSL object should be a connection object and IMO it should NOT change state of other SSL connection objects. SSL_tick() should fail if invoked on stream objects.

Reply:
There is also a section on SSL_tick() behavior below.

Reply:
> The current SSL_tick() API has SSL object as a parameter. That SSL object should be a connection object and IMO it should NOT change state of other SSL connection objects. SSL_tick() should fail if invoked on stream objects.

My concern was not about SSL_tick() being invoked on stream objects. I was describing a scenario where the SSL_tick() is invoked on the connection object, but the error relates to a specific stream and should be reported by the stream object in its next IO call.

Reply:
Do you have any example of such stream specific error during general SSL_tick() processing? Because I do not think there should be any such error.